hawc/apps/epi/api.py

```diff
@@ -87,37 +99,410 @@ def result_heatmap(self, request, pk):
+            if data_key in self.request.data:
+                data_probe = self.request.data[data_key]
+
+                fixed = []
+                for el in data_probe:
+                    if type(el) is str:
+                        try:
+                            criteria = models.Criteria.objects.get(
+                                description=el, assessment_id=assessment_id
+                            )
+                            fixed.append(criteria.id)
+                        except ObjectDoesNotExist:
+                            # Allow creation of criteria as part of the request
+                            criteria_serializer = serializers.CriteriaSerializer(
+                                data={"description": el, "assessment": assessment_id}
+                            )
+                            criteria_serializer.is_valid(raise_exception=True)
+                            criteria = criteria_serializer.save()
+                            fixed.append(criteria.id)
+                    else:
+                        fixed.append(el)
+
+                self.request.data[data_key] = fixed
+
+    def perform_create(self, serializer):
+        super().perform_create(serializer)
+        self.handle_criteria(serializer.instance)
+        self.process_criteria_association(serializer, serializer.instance.id, True)
+
+    def perform_update(self, serializer):
+        super().perform_update(serializer)
+        self.handle_criteria(serializer.instance)
+        self.process_criteria_association(serializer, serializer.instance.id, False)
+
+    @transaction.atomic
+    def update(self, request, *args, **kwargs):
+        return super().update(request, *args, **kwargs)
+
+    @transaction.atomic
+    def create(self, request, *args, **kwargs):
+        # Default behavior except we need to refresh the serializer to get the criteria to show up in the return...
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+
+        instance = self.model.objects.get(id=serializer.data["id"])
+        serializer = self.get_serializer(instance)
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+
+class Exposure(ReadWriteSerializerMixin, PermCheckerMixin, AssessmentEditViewset):
+    perm_checker_key = "study_population"
     assessment_filter_args = "study_population__study__assessment"
     model = models.Exposure
-    serializer_class = serializers.ExposureSerializer
+    read_serializer_class = serializers.ExposureSerializer
+    write_serializer_class = serializers.ExposureWriteSerializer
+
+    def handle_dtxsid(self, request):
+        """
+        Checks supplied dtxsid if present for existence in the database; attempting to create on-the-fly if necessary.
+        """
+        if "dtxsid" in request.data:
+            dtxsid_probe = request.data["dtxsid"]
+            try:
```

Comment:
ideally this logic would be on `assessment.models.DSSTox` overriding the `get_or_create` method, since it's not epi-specific

Reply:
moved, there's still a little left in epi/api.py but the bulk of it (the looking up of the substance and such) is now on DSSTox::save so should be usable outside of epi.